Added support for Slick 3.2.x and cross compilation for Scala 2.12.x

pdolega opened the pull request:
Only uniorn-core is build againt 2.12.x and Slick 3.2 (unicorn-play depends on play-slick and its terribly slow in adapting new versions).

Perhaps also this should target another branch

pdolega commented on .travis.yml:
```diff
@@ -1,9 +1,10 @@
 
 sbt_args: -no-colors
-script: sbt ++$TRAVIS_SCALA_VERSION scalastyle clean coverage test unidoc
+script: sbt ++$TRAVIS_SCALA_VERSION scalastyle clean coverage test unidoc "project unicorn-core" +test
```
no sure how to do it that only `unicorn-core` is cross-built; in current version it builds:
1) everything with 2.11.x once
2) unicorn-core with 2.11 and 2.12  (so 2.11 is built twice).

Kinda inelegant

pdolega commented on .travis.yml:
```diff
@@ -5,5 +5,5 @@ jdk:
 
 sbt_args: -no-colors
-script: sbt ++$TRAVIS_SCALA_VERSION scalastyle clean coverage test unidoc
+script: sbt scalastyle clean coverage test unidoc "project unicorn-core" +test
```
no sure how to do it that only unicorn-core is cross-built; in current version it builds:

1) everything with 2.11.x once
2.1) unicorn-core with 2.11.x (actually it doesn't build it, its already built, it only runs tests second time)
2.2) unicorn-core with 2.12.x

Kinda inelegant (but does the job :/

romanowski replied:
Did you just try to replace 'test' by '+test' in original command?

pdolega replied:
I did, the problem is that it tries to build `unicorn-play` with 2.12 then (which won't work as there is no play-slick for 2.12).

I am actually using `+test` but only narrowed it's scope to `unicorn-play`

pdolega commented on .travis.yml:
```diff
@@ -5,5 +5,5 @@ jdk:
   - oraclejdk8
 
 sbt_args: -no-colors
-script: sbt ++$TRAVIS_SCALA_VERSION scalastyle clean coverage test unidoc
```
One more thing - I am totally puzzled by this: `++$TRAVIS_SCALA_VERSION`. Why would we want Scala version to be picked from travis (while we can have it in sbt independent of OS wide scala version)

romanowski replied:
@Kwestor Can you help here?

Kwestor replied:
No idea, I set this up 3 years ago to have cross-version builds. Feel free to change it as you want.

pdolega replied:
Yeah, I think in the meantime cross building built into sbt has appeared (independent on your actual Scala version)

pdolega replied:
I mean - you don't need to have now multiple Travis envs (with different scala version); it's enough to define it in sbt project

liosedhel commented on unicorn-play/src/main/scala/org/virtuslab/unicorn/UnicornPlay.scala:
```diff
@@ -1,11 +1,11 @@
 import play.api.db.slick.DatabaseConfigProvider
 import play.api.libs.json.Format
-import play.api.mvc.{ PathBindable, QueryStringBindable }
+import play.api.mvc.{PathBindable, QueryStringBindable}
```
Why this braces are formatted differently then in e.g. StringPlayUnicorn?

pdolega replied:
I think it's my mistake - will amend this file (silly me)

pdolega replied:
fixed @liosedhel

liosedhel commented on unicorn-play/src/test/scala/org/virtuslab/unicorn/StringPlayUnicorn.scala:
```diff
@@ -1,8 +1,8 @@
 import com.google.inject.{ Inject, Singleton }
-import slick.backend.DatabaseConfig
-import slick.driver.JdbcProfile
+import org.virtuslab.unicorn.SlickExports.JdbcProfile
```
As I understand, this change in imports is not backward compatible with previous unicorn 3.1.0 version, am I right?

pdolega replied:
Correct - basically this imports change between 3.1 and 3.2 version branches - that's why I aliased them in `SlickExports.scala` (two versions - appropriate compiled per scala version)

pdolega commented:
I squashed commits and fixed @liosedhel comment. 

From my side last thing would be to target appropriate branch. It sure can still be the pointed (as this change builds both versions - 3.1.x and 3.2.x) but maybe you have another opinion on this subject.

@Kwestor @romanowski @liosedhel ^^

liosedhel commented:
For me this branch support new slick 3.2.x version, and has support for 3.1.x slick only for the compatibility reason and only for now. Also this is not backward compatible change (as imports have to be changed). For those two reasons I vote to target new branch `v1.2.x-slick-3.2.x`. @Kwestor @romanowski ?

pdolega commented:
Wait - one clarification. I understood your comment incorrectly: https://github.com/VirtusLab/unicorn/pull/85/files#r94815978

This change is actually backward compatible (from the perspective of Unicorn lib client). Imports in Slick in above comments are not backward compatible. Apologies for misunderstanding.

Basically the point is - anyone who is using Unicorn version before this PR should be able to use Unicorn built after this PR without any change in code whatsoever.

romanowski commented on project/Dependencies.scala:
```diff
@@ -26,4 +26,9 @@ object Dependencies {
 
   val play = mainPlay ++ testPlay
 
+  def pickVersion(scalaVersion: String, ver211: String, ver212: String) = CrossVersion.partialVersion(scalaVersion) match {
```
Sbt has mechanism for that: http://www.scala-sbt.org/0.12.2/docs/Detailed-Topics/Cross-Build.html

pdolega replied:
Right... silly me. Funny thing - I already did something along these lines but forgot now: https://github.com/pdolega/blueprints-scala/blob/master/build.sbt#L35

Getting old... (and will fix :)

pdolega replied:
@romanowski ok, I reconsidered. I am pretty sure this think ( `CrossBuild` ) is used for something else. It basically is used for steering `Scala` version of given library not the library version.

So in essence:
`"com.mycompany" %% "lib" % "3.1.1"` is equal to 
`"com.mycompany" % "lib" % "3.1.1" cross CrossVersion.binary`

All the other variations of above (CrossVersion) are used to drive `XX` part of library: `com.mycompany; lib_XX; 3.1.1`

romanowski replied:
ok, I see. I also get it wrong then.

romanowski commented on project/Settings.scala:
```diff
@@ -1,19 +1,24 @@
     organization := "org.virtuslab",
-    scalaVersion := "2.11.8",
+
+    parallelExecution in Test := false,
```
Why we can't run test in parallel?

pdolega replied:
I just moved the line - it was original here: https://github.com/VirtusLab/unicorn/pull/85/files/e5dd68d039362f30fba476e1b673ba786217c799#diff-6c323177457b467c4590f2132ba7f87aL50

I am guessing that it's because we create / drop same table in every test case.

romanowski commented:
Generally LGTM with few minor questions/comments.

liosedhel commented:
@pdolega I see, it's great to have unicorn fully backward compatible! Still, I would prefer stick to the convention, slick 3.2.x is a new minor version (not just a fix, also we have support for scala 2.12!) and it fits with the new 1.2.x unicorn version. Hopefully, for a long time there will be no new minor version of slick (as right now we have 3.2.0-M2) :) Or, also maybe not a bad idea, we can create slick 3.2.x-unicorn-1.1.x branch and switch to unicorn-1.2.x only when we add some new features (maybe there will be some backward incompatibilities changes needed). What do you think?

pdolega commented:
As for my perspective (which is really not that important here :) - I'd align versions with `Slick` versions. I don't see a point in having evolving changes for e.g. `Slick 3.1.x` (that is - after 3.2 is stable obviously).

Changes between 3.1.x and 3.2.x are minor (from API perspective). This would obviously imply less of work for us (no need to have a version explosion).

2.x is different story. It's totally different than 3.x and it may make sense to keep it for a while (though I am again not sure about anything else that bugfixes).

So my opinion: let's switch to `unicorn 3.2.0` version.

romanowski commented:
LGTM again
